File: app/src/main/java/fr/niavlys/dev/trajets/client/MainActivity.java
package fr.niavlys.dev.trajets.client;

import android.Manifest;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.graphics.Color;
import android.graphics.Typeface;
import android.graphics.drawable.GradientDrawable;
import android.os.Bundle;
import android.telephony.SmsManager;
import android.view.Gravity;
import android.view.View;
import android.widget.Button;
import android.widget.CheckBox;
import android.widget.GridLayout;
import android.widget.LinearLayout;
import android.widget.ScrollView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.app.AppCompatDelegate;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;

import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.function.Supplier;

import fr.niavlys.dev.trajets.config.AppConfig;
import fr.niavlys.dev.trajets.config.ConfigItem;
import fr.niavlys.dev.trajets.config.ConfigRepository;
import fr.niavlys.dev.trajets.config.ContactItem;
import fr.niavlys.dev.trajets.settings.SettingsActivity;

public class MainActivity extends AppCompatActivity {
    private static final int REQUEST_SMS_PERMISSION = 1;

    private ConfigRepository repository;
    private AppConfig config;
    private TextView preview;
    private LinearLayout choicesContainer;
    private LinearLayout recipientsContainer;
    private final Set<String> selectedContactIds = new HashSet<>();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_NO);
        repository = new ConfigRepository(this);
        checkAndRequestSmsPermission();
        buildContent();
    }

    @Override
    protected void onResume() {
        super.onResume();
        if (repository != null) {
            config = repository.load();
            renderRecipients();
            showEmptyChoices();
        }
    }

    private void buildContent() {
        config = repository.load();

        ScrollView scrollView = new ScrollView(this);
        scrollView.setFillViewport(true);
        scrollView.setBackgroundColor(Color.rgb(245, 247, 251));

        LinearLayout root = new LinearLayout(this);
        root.setOrientation(LinearLayout.VERTICAL);
        root.setPadding(dp(18), dp(18), dp(18), dp(24));
        scrollView.addView(root);

        LinearLayout header = new LinearLayout(this);
        header.setGravity(Gravity.CENTER_VERTICAL);
        header.setOrientation(LinearLayout.HORIZONTAL);
        root.addView(header, matchWrap());

        TextView title = title("Trajets");
        LinearLayout.LayoutParams titleParams = new LinearLayout.LayoutParams(0, LinearLayout.LayoutParams.WRAP_CONTENT, 1);
        header.addView(title, titleParams);

        Button settings = secondaryButton("Paramètres");
        settings.setOnClickListener(view -> startActivity(new Intent(this, SettingsActivity.class)));
        header.addView(settings);

        preview = new TextView(this);
        preview.setText("");
        preview.setTextSize(20);
        preview.setTextColor(Color.rgb(21, 32, 43));
        preview.setMinHeight(dp(96));
        preview.setGravity(Gravity.CENTER_VERTICAL);
        preview.setPadding(dp(16), dp(14), dp(16), dp(14));
        preview.setBackground(cardBackground(Color.WHITE));
        LinearLayout.LayoutParams previewParams = matchWrap();
        previewParams.setMargins(0, dp(18), 0, dp(14));
        root.addView(preview, previewParams);

        root.addView(sectionTitle("Construire le message"));
        GridLayout actions = new GridLayout(this);
        actions.setColumnCount(2);
        root.addView(actions, matchWrap());
        addAction(actions, "Je suis à", "Je suis à ", () -> config.getStations());
        addAction(actions, "Je suis dans", "Je suis dans le ", () -> config.getTransports());
        addAction(actions, "Direction", "direction: ", () -> config.getStations());
        addAction(actions, "Je suis avec", "Je suis avec ", this::contactsAsChoices);
        addAction(actions, "Je suis arrivé chez", "Je suis arrivé chez ", this::contactsAsChoices);

        choicesContainer = new LinearLayout(this);
        choicesContainer.setOrientation(LinearLayout.VERTICAL);
        LinearLayout.LayoutParams choicesParams = matchWrap();
        choicesParams.setMargins(0, dp(14), 0, dp(14));
        root.addView(choicesContainer, choicesParams);

        root.addView(sectionTitle("Destinataires SMS"));
        recipientsContainer = new LinearLayout(this);
        recipientsContainer.setOrientation(LinearLayout.VERTICAL);
        recipientsContainer.setBackground(cardBackground(Color.WHITE));
        recipientsContainer.setPadding(dp(12), dp(8), dp(12), dp(8));
        root.addView(recipientsContainer, matchWrap());
        renderRecipients();

        LinearLayout controls = new LinearLayout(this);
        controls.setOrientation(LinearLayout.HORIZONTAL);
        controls.setGravity(Gravity.CENTER);
        LinearLayout.LayoutParams controlsParams = matchWrap();
        controlsParams.setMargins(0, dp(18), 0, 0);
        root.addView(controls, controlsParams);

        Button reset = secondaryButton("Reset");
        reset.setOnClickListener(view -> resetMessage());
        controls.addView(reset, weightedButtonParams());

        Button send = primaryButton("Envoyer");
        send.setOnClickListener(view -> sendMessage());
        controls.addView(send, weightedButtonParams());

        showEmptyChoices();
        setContentView(scrollView);
    }

    private void addAction(GridLayout actions, String label, String prefix, Supplier<List<ConfigItem>> choices) {
        Button button = secondaryButton(label);
        button.setOnClickListener(view -> {
            config = repository.load();
            appendText(prefix);
            renderChoices(choices.get());
        });
        GridLayout.LayoutParams params = new GridLayout.LayoutParams();
        params.width = 0;
        params.height = LinearLayout.LayoutParams.WRAP_CONTENT;
        params.columnSpec = GridLayout.spec(GridLayout.UNDEFINED, 1f);
        params.setMargins(dp(4), dp(4), dp(4), dp(4));
        actions.addView(button, params);
    }

    private void renderChoices(List<ConfigItem> choices) {
        choicesContainer.removeAllViews();
        choicesContainer.addView(sectionTitle("Choix"));

        GridLayout grid = new GridLayout(this);
        grid.setColumnCount(2);
        choicesContainer.addView(grid, matchWrap());

        for (ConfigItem item : choices) {
            Button button = chipButton(item.getLabel());
            button.setOnClickListener(view -> appendText(item.getPhrase()));
            GridLayout.LayoutParams params = new GridLayout.LayoutParams();
            params.width = 0;
            params.height = LinearLayout.LayoutParams.WRAP_CONTENT;
            params.columnSpec = GridLayout.spec(GridLayout.UNDEFINED, 1f);
            params.setMargins(dp(4), dp(4), dp(4), dp(4));
            grid.addView(button, params);
        }
    }

    private void showEmptyChoices() {
        if (choicesContainer == null) {
            return;
        }
        choicesContainer.removeAllViews();
        TextView text = muted("Choisis une action pour afficher les gares, transports ou personnes.");
        text.setGravity(Gravity.CENTER);
        text.setPadding(dp(12), dp(18), dp(12), dp(18));
        text.setBackground(cardBackground(Color.WHITE));
        choicesContainer.addView(text, matchWrap());
    }

    private void renderRecipients() {
        if (recipientsContainer == null || config == null) {
            return;
        }
        recipientsContainer.removeAllViews();
        if (config.getContacts().isEmpty()) {
            recipientsContainer.addView(muted("Aucun numéro configuré."));
            return;
        }
        for (ContactItem contact : config.getContacts()) {
            CheckBox checkBox = new CheckBox(this);
            checkBox.setText(contact.getName() + " · " + contact.getPhone());
            checkBox.setTextSize(16);
            checkBox.setTextColor(Color.rgb(21, 32, 43));
            checkBox.setChecked(selectedContactIds.contains(contact.getId()));
            checkBox.setOnCheckedChangeListener((buttonView, isChecked) -> {
                if (isChecked) {
                    selectedContactIds.add(contact.getId());
                } else {
                    selectedContactIds.remove(contact.getId());
                }
            });
            recipientsContainer.addView(checkBox, matchWrap());
        }
    }

    private List<ConfigItem> contactsAsChoices() {
        java.util.ArrayList<ConfigItem> choices = new java.util.ArrayList<>();
        for (ContactItem contact : config.getContacts()) {
            choices.add(new ConfigItem(contact.getId(), contact.getName(), contact.getName()));
        }
        return choices;
    }

    private void appendText(String text) {
        String before = preview.getText().toString();
        String cleanText = text == null ? "" : text.trim();
        if (cleanText.isEmpty()) {
            return;
        }
        if (!before.isEmpty() && !before.endsWith(" ")) {
            before += " ";
        }
        preview.setText(before + cleanText + " ");
    }

    private void resetMessage() {
        preview.setText("");
        showEmptyChoices();
    }

    private void sendMessage() {
        String message = preview.getText().toString().trim();
        if (message.isEmpty()) {
            Toast.makeText(this, "Le message est vide.", Toast.LENGTH_SHORT).show();
            return;
        }
        if (selectedContactIds.isEmpty()) {
            Toast.makeText(this, "Choisis au moins un destinataire.", Toast.LENGTH_SHORT).show();
            return;
        }
        if (ContextCompat.checkSelfPermission(this, Manifest.permission.SEND_SMS) != PackageManager.PERMISSION_GRANTED) {
            checkAndRequestSmsPermission();
            return;
        }

        SmsManager smsManager = SmsManager.getDefault();
        for (ContactItem contact : config.getContacts()) {
            if (selectedContactIds.contains(contact.getId())) {
                smsManager.sendTextMessage(contact.getPhone(), null, message, null, null);
            }
        }
        Toast.makeText(this, "Message envoyé.", Toast.LENGTH_SHORT).show();
        resetMessage();
    }

    private void checkAndRequestSmsPermission() {
        if (ContextCompat.checkSelfPermission(this, Manifest.permission.SEND_SMS) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.SEND_SMS}, REQUEST_SMS_PERMISSION);
        }
    }

    private TextView title(String text) {
        TextView view = new TextView(this);
        view.setText(text);
        view.setTextSize(30);
        view.setTypeface(Typeface.DEFAULT_BOLD);
        view.setTextColor(Color.rgb(21, 32, 43));
        return view;
    }

    private TextView sectionTitle(String text) {
        TextView view = new TextView(this);
        view.setText(text);
        view.setTextSize(15);
        view.setTypeface(Typeface.DEFAULT_BOLD);
        view.setTextColor(Color.rgb(71, 85, 105));
        view.setPadding(0, dp(12), 0, dp(8));
        return view;
    }

    private TextView muted(String text) {
        TextView view = new TextView(this);
        view.setText(text);
        view.setTextSize(15);
        view.setTextColor(Color.rgb(100, 116, 139));
        return view;
    }

    private Button primaryButton(String text) {
        Button button = baseButton(text);
        button.setTextColor(Color.WHITE);
        button.setBackground(rounded(Color.rgb(37, 99, 235), dp(14)));
        return button;
    }

    private Button secondaryButton(String text) {
        Button button = baseButton(text);
        button.setTextColor(Color.rgb(21, 32, 43));
        button.setBackground(rounded(Color.WHITE, dp(14)));
        return button;
    }

    private Button chipButton(String text) {
        Button button = baseButton(text);
        button.setTextColor(Color.rgb(30, 64, 175));
        button.setBackground(rounded(Color.rgb(219, 234, 254), dp(14)));
        return button;
    }

    private Button baseButton(String text) {
        Button button = new Button(this);
        button.setText(text);
        button.setAllCaps(false);
        button.setTextSize(15);
        button.setMinHeight(dp(48));
        button.setPadding(dp(12), dp(8), dp(12), dp(8));
        return button;
    }

    private GradientDrawable cardBackground(int color) {
        GradientDrawable drawable = rounded(color, dp(18));
        drawable.setStroke(dp(1), Color.rgb(226, 232, 240));
        return drawable;
    }

    private GradientDrawable rounded(int color, int radius) {
        GradientDrawable drawable = new GradientDrawable();
        drawable.setColor(color);
        drawable.setCornerRadius(radius);
        return drawable;
    }

    private LinearLayout.LayoutParams matchWrap() {
        return new LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT,
                LinearLayout.LayoutParams.WRAP_CONTENT
        );
    }

    private LinearLayout.LayoutParams weightedButtonParams() {
        LinearLayout.LayoutParams params = new LinearLayout.LayoutParams(0, LinearLayout.LayoutParams.WRAP_CONTENT, 1);
        params.setMargins(dp(4), 0, dp(4), 0);
        return params;
    }

    private int dp(int value) {
        return (int) (value * getResources().getDisplayMetrics().density);
    }

    @Deprecated
    public static void addPrevisualisation(String text) {
    }

    @Deprecated
    public static void resetPrevisualisation() {
    }

    @Deprecated
    public static Button getPrevious() {
        return null;
    }

    @Deprecated
    public static Button getNext() {
        return null;
    }
}

File: app/src/main/java/fr/niavlys/dev/trajets/client/MainActivity.kt
package fr.niavlys.dev.trajets.client

import android.Manifest
import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.Color
import android.graphics.Typeface
import android.graphics.drawable.GradientDrawable
import android.os.Bundle
import android.telephony.SmsManager
import android.view.Gravity
import android.view.View
import android.widget.Button
import android.widget.CheckBox
import android.widget.GridLayout
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.ScrollView
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.app.AppCompatDelegate
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import fr.niavlys.dev.trajets.R
import fr.niavlys.dev.trajets.config.AppConfig
import fr.niavlys.dev.trajets.config.ConfigItem
import fr.niavlys.dev.trajets.config.ConfigRepository
import fr.niavlys.dev.trajets.config.ContactItem
import fr.niavlys.dev.trajets.settings.SettingsActivity

class MainActivity : AppCompatActivity() {
    private lateinit var repository: ConfigRepository
    private lateinit var config: AppConfig
    private lateinit var preview: TextView
    private lateinit var choicesContainer: LinearLayout
    private lateinit var recipientsContainer: LinearLayout

    private val selectedContactIds = mutableSetOf<String>()
    private val appBackground = Color.rgb(246, 248, 252)
    private val surface = Color.WHITE
    private val textPrimary = Color.rgb(15, 23, 42)
    private val textSecondary = Color.rgb(100, 116, 139)
    private val border = Color.rgb(226, 232, 240)
    private val primary = Color.rgb(37, 99, 235)
    private val primarySoft = Color.rgb(219, 234, 254)
    private val success = Color.rgb(16, 185, 129)
    private val danger = Color.rgb(220, 38, 38)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_NO)
        repository = ConfigRepository(this)
        checkAndRequestSmsPermission()
        buildScreen()
    }

    override fun onResume() {
        super.onResume()
        if (::repository.isInitialized && ::recipientsContainer.isInitialized) {
            config = repository.load()
            selectedContactIds.retainAll(config.contacts.map { it.id }.toSet())
            renderRecipients()
            showEmptyChoices()
        }
    }

    private fun buildScreen() {
        config = repository.load()

        val scrollView = ScrollView(this).apply {
            setFillViewport(true)
            setBackgroundColor(appBackground)
        }

        val root = LinearLayout(this).apply {
            orientation = LinearLayout.VERTICAL
            setPadding(dp(18), dp(18), dp(18), dp(24))
        }
        scrollView.addView(root, matchWrap())

        root.addView(header(), matchWrap())
        root.addView(previewCard(), marginLayout(top = 18, bottom = 18))
        root.addView(sectionTitle("Construire le message"))
        root.addView(actionGrid(), matchWrap())

        choicesContainer = LinearLayout(this).apply { orientation = LinearLayout.VERTICAL }
        root.addView(choicesContainer, marginLayout(top = 14, bottom = 18))

        root.addView(sectionTitle("Destinataires SMS"))
        recipientsContainer = LinearLayout(this).apply {
            orientation = LinearLayout.VERTICAL
            setPadding(dp(12), dp(8), dp(12), dp(8))
            background = cardBackground(surface)
        }
        root.addView(recipientsContainer, matchWrap())
        renderRecipients()

        root.addView(bottomControls(), marginLayout(top = 18))
        showEmptyChoices()
        setContentView(scrollView)
    }

    private fun header(): LinearLayout = LinearLayout(this).apply {
        orientation = LinearLayout.HORIZONTAL
        gravity = Gravity.CENTER_VERTICAL

        val logo = ImageView(this@MainActivity).apply {
            setImageResource(R.drawable.ic_launcher_foreground)
            background = rounded(primarySoft, 18)
            setPadding(dp(4), dp(4), dp(4), dp(4))
        }
        addView(logo, LinearLayout.LayoutParams(dp(52), dp(52)))

        val titleBlock = LinearLayout(this@MainActivity).apply {
            orientation = LinearLayout.VERTICAL
            setPadding(dp(12), 0, 0, 0)
            addView(text("Trajets", 30f, textPrimary, Typeface.BOLD))
            addView(text("Messages rapides et configurables", 14f, textSecondary))
        }
        addView(titleBlock, LinearLayout.LayoutParams(0, LinearLayout.LayoutParams.WRAP_CONTENT, 1f))

        addView(secondaryButton("Paramètres").apply {
            setOnClickListener { startActivity(Intent(this@MainActivity, SettingsActivity::class.java)) }
        })
    }

    private fun previewCard(): LinearLayout = LinearLayout(this).apply {
        orientation = LinearLayout.VERTICAL
        setPadding(dp(16), dp(14), dp(16), dp(14))
        background = cardBackground(surface)

        addView(text("Message", 13f, textSecondary, Typeface.BOLD))
        preview = text("", 20f, textPrimary).apply {
            minHeight = dp(82)
            gravity = Gravity.CENTER_VERTICAL
            text = "Ton message apparaîtra ici"
            setTextColor(textSecondary)
        }
        addView(preview, matchWrap())
    }

    private fun actionGrid(): GridLayout = GridLayout(this).apply {
        columnCount = 2
        addAction("Je suis à", "Je suis à", { config.stations })
        addAction("Je suis dans", "Je suis dans le", { config.transports })
        addAction("Direction", "direction:", { config.stations })
        addAction("Je suis avec", "Je suis avec", ::contactsAsChoices)
        addAction("Je suis arrivé chez", "Je suis arrivé chez", ::contactsAsChoices)
    }

    private fun GridLayout.addAction(label: String, prefix: String, choices: () -> List<ConfigItem>) {
        val button = secondaryButton(label).apply {
            setOnClickListener {
                config = repository.load()
                appendSegment(prefix)
                renderChoices(choices())
            }
        }
        addView(button, gridParams())
    }

    private fun renderChoices(choices: List<ConfigItem>) {
        choicesContainer.removeAllViews()
        choicesContainer.addView(sectionTitle("Choix"))

        if (choices.isEmpty()) {
            choicesContainer.addView(emptyCard("Aucun élément configuré. Ajoute-le depuis Paramètres."))
            return
        }

        val grid = GridLayout(this).apply { columnCount = 2 }
        choicesContainer.addView(grid, matchWrap())
        choices.forEach { item ->
            val chip = chipButton(item.label).apply {
                setOnClickListener { appendSegment(item.phrase) }
            }
            grid.addView(chip, gridParams())
        }
    }

    private fun showEmptyChoices() {
        choicesContainer.removeAllViews()
        choicesContainer.addView(emptyCard("Choisis une action pour afficher les gares, transports ou personnes."))
    }

    private fun renderRecipients() {
        recipientsContainer.removeAllViews()
        if (config.contacts.isEmpty()) {
            recipientsContainer.addView(muted("Aucun numéro configuré."))
            return
        }

        config.contacts.forEach { contact ->
            val checkBox = CheckBox(this).apply {
                text = "${contact.name}  ·  ${contact.phone}"
                textSize = 16f
                setTextColor(textPrimary)
                isChecked = selectedContactIds.contains(contact.id)
                setOnCheckedChangeListener { _, checked ->
                    if (checked) selectedContactIds += contact.id else selectedContactIds -= contact.id
                }
            }
            recipientsContainer.addView(checkBox, matchWrap())
        }
    }

    private fun bottomControls(): LinearLayout = LinearLayout(this).apply {
        orientation = LinearLayout.HORIZONTAL
        gravity = Gravity.CENTER

        addView(secondaryButton("Effacer").apply { setOnClickListener { resetMessage() } }, weightedButtonParams())
        addView(primaryButton("Envoyer").apply { setOnClickListener { sendMessage() } }, weightedButtonParams())
    }

    private fun contactsAsChoices(): List<ConfigItem> =
        config.contacts.map { ConfigItem(it.id, it.name, it.name) }

    private fun appendSegment(segment: String) {
        val value = segment.trim()
        if (value.isEmpty()) return

        val current = preview.text.toString()
        val base = if (current == "Ton message apparaîtra ici") "" else current
        val separator = if (base.isNotBlank() && !base.endsWith(" ")) " " else ""
        preview.setTextColor(textPrimary)
        preview.text = "$base$separator$value "
    }

    private fun resetMessage() {
        preview.setTextColor(textSecondary)
        preview.text = "Ton message apparaîtra ici"
        showEmptyChoices()
    }

    private fun sendMessage() {
        val message = preview.text.toString().trim()
        if (message.isEmpty() || message == "Ton message apparaîtra ici") {
            toast("Le message est vide.")
            return
        }
        if (selectedContactIds.isEmpty()) {
            toast("Choisis au moins un destinataire.")
            return
        }
        if (ContextCompat.checkSelfPermission(this, Manifest.permission.SEND_SMS) != PackageManager.PERMISSION_GRANTED) {
            checkAndRequestSmsPermission()
            toast("Autorise l'envoi de SMS pour continuer.")
            return
        }

        val smsManager = SmsManager.getDefault()
        config.contacts
            .filter { selectedContactIds.contains(it.id) }
            .forEach { smsManager.sendTextMessage(it.phone, null, message, null, null) }

        toast("Message envoyé.")
        resetMessage()
    }

    private fun checkAndRequestSmsPermission() {
        if (ContextCompat.checkSelfPermission(this, Manifest.permission.SEND_SMS) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(this, arrayOf(Manifest.permission.SEND_SMS), REQUEST_SMS_PERMISSION)
        }
    }

    private fun toast(message: String) {
        Toast.makeText(this, message, Toast.LENGTH_SHORT).show()
    }

    private fun primaryButton(label: String): Button = baseButton(label).apply {
        setTextColor(Color.WHITE)
        background = rounded(primary, 14)
    }

    private fun secondaryButton(label: String): Button = baseButton(label).apply {
        setTextColor(textPrimary)
        background = outlined(surface, 14)
    }

    private fun chipButton(label: String): Button = baseButton(label).apply {
        setTextColor(Color.rgb(30, 64, 175))
        background = rounded(primarySoft, 14)
    }

    private fun baseButton(label: String): Button = Button(this).apply {
        text = label
        isAllCaps = false
        textSize = 15f
        minHeight = dp(48)
        setPadding(dp(12), dp(8), dp(12), dp(8))
    }

    private fun sectionTitle(label: String): TextView = text(label, 15f, Color.rgb(71, 85, 105), Typeface.BOLD).apply {
        setPadding(0, dp(8), 0, dp(8))
    }

    private fun muted(label: String): TextView = text(label, 15f, textSecondary)

    private fun emptyCard(label: String): TextView = muted(label).apply {
        gravity = Gravity.CENTER
        setPadding(dp(12), dp(18), dp(12), dp(18))
        background = cardBackground(surface)
    }

    private fun text(label: String, size: Float, color: Int, style: Int = Typeface.NORMAL): TextView =
        TextView(this).apply {
            text = label
            textSize = size
            setTextColor(color)
            typeface = Typeface.create(Typeface.DEFAULT, style)
        }

    private fun gridParams(): GridLayout.LayoutParams = GridLayout.LayoutParams().apply {
        width = 0
        height = LinearLayout.LayoutParams.WRAP_CONTENT
        columnSpec = GridLayout.spec(GridLayout.UNDEFINED, 1f)
        setMargins(dp(4), dp(4), dp(4), dp(4))
    }

    private fun matchWrap(): LinearLayout.LayoutParams =
        LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.WRAP_CONTENT)

    private fun marginLayout(top: Int = 0, bottom: Int = 0): LinearLayout.LayoutParams =
        matchWrap().apply { setMargins(0, dp(top), 0, dp(bottom)) }

    private fun weightedButtonParams(): LinearLayout.LayoutParams =
        LinearLayout.LayoutParams(0, LinearLayout.LayoutParams.WRAP_CONTENT, 1f).apply {
            setMargins(dp(4), 0, dp(4), 0)
        }

    private fun cardBackground(color: Int): GradientDrawable = outlined(color, 18)

    private fun outlined(color: Int, radius: Int): GradientDrawable =
        rounded(color, radius).apply { setStroke(dp(1), border) }

    private fun rounded(color: Int, radius: Int): GradientDrawable = GradientDrawable().apply {
        setColor(color)
        cornerRadius = dp(radius).toFloat()
    }

    private fun dp(value: Int): Int = (value * resources.displayMetrics.density).toInt()

    companion object {
        private const val REQUEST_SMS_PERMISSION = 1
    }
}
